src/python/WMCore/MicroService/Unified/MSRuleCleaner.py

```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+        already worked on the workflow as a mask and applies this mask over
+        the set of flags in the 'CleanupStatus' field and then reduces the
+        result to a single bool value
+        """
+        # NOTE: This is one of the few functions taking a workflow as an argument
+        #       but returning a bool, since it is an auxiliary function and is not
+        #       supposed to be called as a standalone function in a pipeline.
+        # NOTE: `all([]) == True`, ergo all the 'rejected' && 'aborted-completed' workflows
+        #       are also counted as properly cleaned and can trigger archival later
+        cleanStatus = all([wflow['CleanupStatus'][key]
+                           for key in wflow['PlineMarkers']
+                           if key in wflow['CleanupStatus'].keys()])
+        return cleanStatus
+
+    def setClean(self, wflow):
+        """
+        A function to set the 'IsClean' flag based on the status from all the
+        pipelines which have worked on the workflow (and have put their markers
+        in the 'PlineMarkers' list)
+        :param  wflow:      A MSRuleCleaner workflow representation
+        :return wflow:
+        """
+        wflow['IsClean'] = self._checkClean(wflow)
+        return wflow
+
+    def archive(self, wflow):
+        """
+        Move the workflow to the proper archived status after checking
+        the full cleanup status
+        :param  wflow:      A MSRuleCleaner workflow representation
+        :param  archStatus: Target status to transition after archival
+        :return wflow:
+        """
+        # NOTE: check allowed status transitions with:
+        #       https://github.com/dmwm/WMCore/blob/5961d2229b1e548e58259c06af154f33bce36c68/src/python/WMCore/ReqMgr/DataStructs/RequestStatus.py#L171
+        if not wflow['IsClean']:
+            msg = "Not properly cleaned workflow: %s" % wflow['RequestName']
+            raise MSRuleCleanerArchival(msg)
+
+        # Check the available status transitions before we decide the final status
+        targetStatusList = RequestStatus.REQUEST_STATE_TRANSITION.get(wflow['RequestStatus'], [])
+        self.logger.info("targetStatusList: %s", targetStatusList)
+        return wflow
+
+    def getMSOutputTransferInfo(self, wflow):
+        """
+        Fetches the transfer information from the MSOutput REST interface for
+        the given workflow.
+        :param  wflow:   A MSRuleCleaner workflow representation
+        :return wflow:
+        """
+        headers = {'Accept': 'application/json'}
+        params = {}
+        url = '%s/data/info?request=%s' % (self.msConfig['MSOutputUrl'],
+                                           wflow['RequestName'])
+        mgr = RequestHandler()
+        res = mgr.getdata(url, params=params, headers=headers, ckey=ckey(), cert=cert())
+        data = json.loads(res)['result'][0]
+        wflow['TransferInfo'] = data['transferDoc']
```

Comment:
I made a comment up in the data structure. I see no need to store the whole (even though it's not big) transfer info data structure here. In the end, we only care about the final status of the MSOutput output data placement.

Reply:
Done